### PR TITLE
ZDC unpacker, cleaner commit

### PR DESCRIPTION
**Pull request by cferraio**
Hi all,

I'm started a new commit with the ZDC unpacker that is much cleaner. Hopefully this helps.

Chris

### CODE REVIEW

**Comment by mmusich** [comment]
-1
@cferraio you're still proposing to remove the changes I recently introduce in #12491. A further rebase is needed

**Comment by mmusich** [comment]
hold

**Comment by cferraio** [comment]
Hi @mmusich, how do I do this? I thought I had taken the most recent release.

Sorry...first time going through this progress...

**Comment by abdoulline** [comment]
The most recent IB is usually as good way to proceed...
  scramv1 list -a CMSSW  |  grep 7_5_X
->  CMSSW_7_5_X_2015-11-19-1100

**Comment by cferraio** [comment]
Ok, that's the one I used, however it looks like i DID add in AlCaReco...py
"git add Configuration/StandardSequences/python/AlCaRecoStreamsHeavyIons_cff.py"

I think I understand what happened now, I copied over a folder instead of an individual file, and then committed what git picked up as modified. If I change that file to what Marco committed, then recommit, would that work, or should I start over again and be more careful when copying files?

Thanks,
Chris

**Comment by mmusich** [comment]
@cferraio you can edit the file and remove the changes you didn't mean to include. After that you can commit again with the --amend option (that will make your history look good).

**Comment by cferraio** [comment]
hi @mmusich, i'm having some issues. i made the change and committed with amend, but i receive an error message when i push:
git push my-cmssw ZdcUnpacker_HI75X
Enter passphrase for key '/afs/cern.ch/user/c/cferraio/.ssh/id_rsa': 
To git@github.com:cferraio/cmssw.git
 ! [rejected]        ZdcUnpacker_HI75X -> ZdcUnpacker_HI75X (non-fast-forward)
error: failed to push some refs to 'git@github.com:cferraio/cmssw.git'
hint: Updates were rejected because the tip of your current branch is behind
hint: its remote counterpart. Merge the remote changes (e.g. 'git pull')
hint: before pushing again.
hint: See the 'Note about fast-forwards' in 'git push --help' for details.

thanks again...

**Comment by mmusich** [comment]
`git push my-cmssw +HEAD:ZdcUnpacker_HI75X`

**Comment by cferraio** [comment]
great, that did it, thanks!

fingers crossed everything looks good...

**Comment by mmusich** [comment]
there is still one line difference, but should be harmless enough :)

**Comment by mmusich** [comment]
unhold

**Comment by slava77** [comment]
unless I'm missing something, this is still working only up to digi. The issue with rechits is still to be addressed, right?

**Comment by cferraio** [comment]
Hi Slava, 

Sorry, maybe I was confused, I thought this was something we were to work on after we got this up and running and into the digi level. Is it necessary to fix through reco at this stage?

**Comment by slava77** [comment]
Chris,
up to your last message and the PR updates just recently I was anticipating the local reco is going to be updated.
High hopes, I guess.

**Comment by slava77** [comment]
considering that all past setups are effectively working because the zdc unpacking is still also running in hcal and the current "old" local reco reads from there and will get the right input for all old data, the current state is on an incremental track of making new data to be processed correctly.

**Comment by cferraio** [comment]
Ah, I see. Sorry, we're trying to get it in one step at a time. Unfortunately the resources dedicated to the software are rather limited (i.e. it's just me). As soon as the Raw->Digi is done, I'll be working right on the Reco. The Raw->Digi is essential for us because that much is enough to allow the Zdc group to perform some basics tests, make sure the gain on our PMTs is set correctly, etc. So, shoe horning the working Raw->Digi code into cmssw as soon as we can is the priority, then we'll worry about reco...

That said, old data should run through flawlessly, new data will need some hopefully minor edits.

**Comment by slava77** [comment]
Chris,
what are the chances that you will need to update the unpacker? Is it fully debugged now with the new data?
I'm trying to understand if we gain much and you can't just use RAW input data for the DPG studies and the ZDC detector performance analysis.

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by cferraio** [comment]
for this PR it should be fully debugged, i still have to run it through 76X and 80X and get it up and running there. from what i understand (and this may not be true, but it's the information that was passed on to me) is that we might not have access to the RAW data. i can look into this further.

File: EventFilter/CastorRawToDigi/src/ZdcUnpacker.cc [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+    // get pointers
+    htr.dataPointers(&daq_first,&daq_last,&tp_first,&tp_last);
+    unsigned int smid=htr.getSubmodule();
+    int htr_tb=smid&0x1;
+    int htr_slot=(smid>>1)&0x1F;
+    int htr_cr=(smid>>6)&0x1F;
+
+    //////////////////////////////////////////////
+    qie_begin=(const HcalQIESample*)daq_first;
+    qie_end=(HcalQIESample*)(daq_last+1); // one beyond last..
```
Comment (slava77) [inline EventFilter/CastorRawToDigi/src/ZdcUnpacker.cc]: this generates a complaint from the static analyzer as well 
(you can see it in the link next to "Tested at: " after automated test report is produced, they come out in "Static checks outputs")
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-12497/9866/llvm-analysis/report-5dee98.html#EndPath

the code in fact doesn't use these casted non-consts, no harm.
Still, better fix it.

**Comment by slava77** [comment]
I'm confused now: I was expecting that after we move to castor raw2digi, only one ZDC FED will be unpacked (the 693).
[this picture doesn't quite match the slides from today. What I see in the code now is that both FED 693 and FED 722 are unpacked in one place]

What's FED 722? it's the old ZDC VME FED index, when it was "attached" to HF, right?

I have a feeling I understand even less now.

Let me start with "basic" questions:
- can the hcal raw2digi unpack new ZDC data if it was coming from the FED 722? (was it a mere FED number change or was there really a change in the format?)
- if it's just a FED numbering, can't you just add FED 693 to https://github.com/cms-sw/cmssw/blob/CMSSW_7_5_6/EventFilter/HcalRawToDigi/plugins/HcalRawToDigi.cc#L39 and be done?

Looking at the HcalUnpacker and HcalRawToDigi, it wouldn't care which VME FED number it is as long as it's in the list to try.
@abdoulline ?

**Comment by slava77** [comment]
sorry, I meant
- can the hcal raw2digi unpack new ZDC data if it was coming from the FED 693? (was it a mere FED number change or was there really a change in the format?)

**Comment by abdoulline** [comment]
Salava, 
it was HCAL decision to start "phasing out" ZDC from HCAL, as new ZDC is sitting in CASTOR crate/FED
http://cmslxr.fnal.gov/source/DataFormats/FEDRawData/interface/FEDNumbering.h

ZDC entries were _removed_ from HCAL emap since April. This was no-return point unpacking-wise.

File: EventFilter/CastorRawToDigi/src/ZdcUnpacker.cc (hunk 1 above)
Comment (cferraio) [inline EventFilter/CastorRawToDigi/src/ZdcUnpacker.cc]: Ah, ok, I missed the const on the second line. Fixed.

**Comment by cferraio** [comment]
Hi Slava,

Unfortunately Hcal has kicked out the Zdc :-).

That's correct, 722 is the Fed used for old datasets, and I'm using it to distinguish between pre- and post- April data sets. As Salavat mentioned, Hcal can't simply unpack Zdc with a change of fed id for run 2 data. I've set up the code so that when it has info in fed 722 it uses different electronics and detectors ids than when it unpacks info in fed 693.

**Comment by slava77** [comment]
@cmsbuild please test
[will just check up to compilation]

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/12497 1b2649d
- code changes are OK for this emergency 75X situation (and I'd prefer not to forward port and instead to solve this appropriately; even better still I'd rather see the ZDC unpacking go back to hcalDigis where it was fitting well and will still fit OK from the code perspective)
- jenkins tests pass and comparisons with baseline show no relevant differences
- local tests with CMSSW_7_5_6 and run2 pp data and also recent runs (262163) 
    - show only a change in HcalUnpackerReport castorDigis errorFree among the monitored quantities (expected)
    - zdc digis are not empty with a manual inspection in 262163

For post-75X perspective. If I understood the plans correctly, the 722 and 693 FED unpacking (now both in castorDigis) will be moved out to a separate zdcRawToDigi and also outputs of it will be picked up by the local reconstruction.

**Comment by slava77** [comment]
merge

compilation is OK from the console in jenkins

**Pull request merged**